### ISSUE TITLE
Large bundle size of Office Fabric UI React web parts in SPFx

### ISSUE THREAD

**Issue by nishantverma-nagarro** [OfficeDev/office-ui-fabric-react#9492]
Hi,
 
I am creating web parts that will go on a same page. The web part count can be from 10-15. I am using SPFx with React and have added Office Fabric UI React to the web parts as recommended in MSDN. Now when I bundle the web parts the size of the bundled files are large. For example, the smallest file is of 104 KB. So now if add 10-15 web parts on the same page, my page will be very slow.
I need to reduce the bundle size of the files. Hence I followed this link to use office fabric react an external link to the web parts. I added the following code in the externals in config.json:

"office-ui-fabric-react": {
"path": "https://unpkg.com/office-ui-fabric-react@6.143.0/dist/office-ui-fabric-react.min.js",
"globalName": "Fabric"
}

Now when the files are bundled, the size reduces from 104 KB to 45 KB. This was a good reduction.
I deployed the web parts to SharePoint, but the office fabric did not load. The web parts were throwing following error: 
 Uncaught ReferenceError: React is not defined
So in order to remove the error I thought of adding the react CDN path to the page. I did this on classic site using Script Editor web part and the web part worked!
 
To add the same CDN path to the Modern UI we created a SPFx extension and installed it to SharePoint. But when I added the web parts to the page they threw the same error: "React is not defined". I tried adding the CDN paths first in the page head so that the CDN files load before the fabric file. But this did not succeed.
 
I then added the CDN path in the external property of config.json so that they may load in sequence. But on doing so the bundling process failed and the console threw the following error:

Error - [copy-assets] External "react" is specified both as a linked package in the node_modules dependency tree and as an external in in the config.json file. Remove one of the external references. Only one external reference type is allowed per
external package.

I also installed webpack 4 but that too did not bring the bundle size down.

So now my challenge is to reduce the bundle size of the Office Fabric UI React Web Part files. Has anyone faced the same issue? Is there any way to reduce the bundle size of the web parts?
 
Regards,
Nishant

**Comment by aftab-hassan** [comment]
@nishantverma-nagarro Thanks for opening an issue. Having 10-15 web parts, with each file being ~100 kB in size would be detrimental to page load time. 

I shall discuss this with the team, and possibly with someone  from SPFx and reply back here, later today.

**Comment by aftab-hassan** [comment]
@nishantverma-nagarro 

Spoke with the team and we have a couple different options that could help with reducing the bundle size.

You could use `react` as an external global dependency. [Here](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/basics/add-an-external-library#load-a-library-that-has-a-dependency-on-another-library
) is some guidance. 
You could try adding a field in the `config.json` at solution level, similar to the following
```
{
    "react": {
        "path": "CDN path to react",
        "globalName": "react"
    },
    "office-ui-fabric-react": {
        "path": "CDN path to office-ui-fabric-react",
        "globalName": "office-ui-fabric-react",
        "globalDependencies": ["react"]
    }
}
```

The above, however, was not designed for this purpose, and is not the recommended way to use `office-ui-fabric-react` + `SPFx`. 

The recommended way would be to create an SPFx Library component, which will bundle the `office-ui-fabric-react` components you have dependencies on, in to the library. The solution can then take a dependency on the SPFx Library component you created, of which there would only be one instance. This is currently in preview. [Here](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/library-component-overview) is some guidance.  

Please let us know if this helps.

**Comment by nishantverma-nagarro** [comment]
Hi,

The first option looks tricky because I have to create interface that maps all the functions in both Office fabric UI React and React itself. Hope this is what you meant. And also it is not recommended and hence it will not go down well with our clients.

For the second option as well we cannot use it because it is still in preview and so we cannot use it in production web parts. We will be waiting for the actual release so that we can take advantage of the libraries. Hope you will have documentations as to how we can use office fabric ui react in the libraries in order to reduce the bundle size. I hope you understand that this is a major issue for us.
 
We are concerned because when we create a react web part using React, the non minimized file size is about 40KB which will be minimized to even smaller file size. But if you add a fabric component like a single Icon on the same web part the size jumps to 340KB for non minimized file and 105KB for minimized file. Now that is a huge issue for the performance of a page when we add a dozen web parts on the same page.

Regards,
Nishant

**Comment by ps-from-md** [comment]
I have been struggling with a similar issue with bundle sizes.  My assumption was that if I correctly used imports (import {Button} from 'office-fabric-ui-react') that when webpack ran, it would only include the items I referenced.  However, each time I run webpack analyzer I can see that many components are included that are not referenced and my bundle size is pushing 3 MB (yes, MB).

So, I configured webpack to splitchunks and pushed all into a commons.bundle.js file, set up the aggressive merge plugins as well as uglyfyjs and then did another analysis.  The office-fabric-ui portion of the commons bundle is still 2 MB!

Now, with that being said... I dont honestly think that this is directly an issue with this package, more so, an issue with how webpack is configured as well as how imports are referenced...  I state this because when i review the  other packages in the commons file they are large as well, though none exceeding 500KB.

```
optimization: {
        minimizer: [new UglifyJsPlugin()],
        splitChunks: {
            cacheGroups: {
                commons: {
                    name: 'commons',
                    chunks: 'all'
                }
            }
        }
    },
    module: {
        rules: [
            { test: /\.(js|jsx)$/, exclude: /node_modules/, loader: 'babel-loader' },
			{ test: /\.(png|woff|woff2|eot|ttf|svg)$/, loader: 'url-loader?limit=100000' },
            { test: /\.scss$/, use: [{ loader: 'style-loader' }, { loader: 'css-loader' },{ loader: 'sass-loader' }]}, 
            { test: /\.tsx?$/, loader: 'babel-loader', exclude: [/node_modules/,nodeModulesPath],
              query: {
                cacheDirectory: false,
                plugins: [
                    require.resolve('babel-plugin-transform-decorators-legacy'),
                ]}},
            { test: /\.tsx$/, exclude: [/node_modules/,nodeModulesPath], loader: "ts-loader?configFile=tsconfig.json" }
        ]
    }
```

![image](https://user-images.githubusercontent.com/13452627/59942909-e8ffde80-942e-11e9-8f21-41326766b854.png)

So @aftab-hassan if you have suggestions as to what could be causing this bloat, please us know.  I am in a unique situation where we cannot reference CDNs due to an air gap network, so we must bundle correctly.

thanks!

**Comment by aftab-hassan** [comment]
> We will be waiting for the actual release so that we can take advantage of the libraries. Hope you will have documentations as to how we can use office fabric ui react in the libraries in order to reduce the bundle size. I hope you understand that this is a major issue for us.

@nishantverma-nagarro We understand this is a major issue and affects page load time and performance.

The `SPFx Library component` will be `GA` in the near future. Please try using it when `GA` and let us know if it was helpful in your scenario. Also, please feel free to post your questions with an `SPFx` label on `stack overflow` as well since this page tracks mainly `Fabric` issues, and you might be able to get a quicker response there.

**Comment by aftab-hassan** [comment]
@ps-from-md thanks for commenting on this issue. I don't have an answer to your question right away, but will come back once I have something to share.

**Comment by ps-from-md** [comment]
@aftab-hassan @nishantverma-nagarro 

So, I spent a couple hours on this last Friday trying to figure out there was anything in our code-base that could be causing issues.  I tried many different changes, but nothing seemed to work... except one.

When I updated the imports to reference the items in the lib, the package shrunk significantly.  In the WebStorm IDE, when you have multiple imports that referencing the same path, it often refactors it for you so that all items are in the same line, like this:

`import {PrimaryButton, MessageBarType, TextField, MessageBar, Stack, Separator, Icon} from "office-ui-fabric-react";`

When I went through all the classes and removed the single line and pointed each to the specific control,  it would appear the tree-shaking worked and the excess code was not imported, like this:

```
import {MessageBarType, MessageBar} from "office-ui-fabric-react/lib/MessageBar";
import {Stack} from "office-ui-fabric-react/lib/Stack";
import {Separator} from "office-ui-fabric-react/lib/Separator";
import {TextField} from "office-ui-fabric-react/lib/TextField";
import {PrimaryButton} from "office-ui-fabric-react/lib/Button";
import {Icon} from "office-ui-fabric-react/lib/Icon";
```

So, with that being said, if anyone is still having issues, i would try the above change and see if it helps.
